graph/loaders/comments.js

```diff
@@ -1,85 +1,134 @@
+ * @param {Array<String>} asset_ids   the ids of assets for which there are
+ *                                    comments that we want to get
  */
-const genCommentsByAssetID = (context, ids) => {
-  return CommentModel.find({
-    asset_id: {
-      $in: ids
+const getCountsByAssetID = (context, asset_ids) => {
+  return CommentModel.aggregate([
+    {
+      $match: {
+        asset_id: {
+          $in: asset_ids
+        },
+        status: {
+          $in: [null, 'ACCEPTED']
+        },
+        parent_id: null
+      }
     },
-    parent_id: null,
-    status: {
-      $in: [null, 'ACCEPTED']
+    {
+      $group: {
+        _id: '$asset_id',
+        count: {
+          $sum: 1
+        }
+      }
     }
-  })
-  .sort({created_at: -1})
-  .then(util.arrayJoinBy(ids, 'asset_id'));
+  ])
+  .then(util.singleJoinBy(asset_ids, '_id'))
+  .then((results) => results.map((result) => result ? result.count : 0));
 };
 
 /**
- * Retrieves comments by an array of parent ids, results are returned in
- * chronological order.
- * @param {Array} ids array of ids to lookup
+ * Returns the comment count for all comments that are public based on their
+ * parent ids.
+ * @param {Object}        context     graph context
+ * @param {Array<String>} parent_ids  the ids of parents for which there are
+ *                                    comments that we want to get
  */
-const genCommentsByParentID = (context, ids) => {
-  return CommentModel.find({
-    parent_id: {
-      $in: ids
+const getCountsByParentID = (context, parent_ids) => {
+  return CommentModel.aggregate([
+    {
+      $match: {
+        parent_id: {
+          $in: parent_ids
+        },
+        status: {
+          $in: [null, 'ACCEPTED']
+        }
+      }
     },
-    status: {
-      $in: [null, 'ACCEPTED']
+    {
+      $group: {
+        _id: '$parent_id',
+        count: {
+          $sum: 1
+        }
+      }
     }
-  })
-  .sort({created_at: 1})
-  .then(util.arrayJoinBy(ids, 'parent_id'));
+  ])
+  .then(util.singleJoinBy(parent_ids, '_id'))
+  .then((results) => results.map((result) => result ? result.count : 0));
 };
 
-const getCommentsByStatusAndAssetID = (context, {status = null, asset_id = null}) => {
+/**
+ * Retrieves comments based on the passed in query that is filtered by the
+ * current used passed in via the context.
+ * @param  {Object} context   graph context
+ * @param  {Object} query     query terms to apply to the comments query
+ */
+const getCommentsByQuery = ({user}, {ids, statuses, asset_id, parent_id, limit, cursor, sort}) => {
+  let comments = CommentModel.find();
 
-  // TODO: remove when we move the enum over to the uppercase.
-  if (status) {
-    status = status.toLowerCase();
+  // Only administrators can search for comments with statuses that are not
+  // `null`, or `'ACCEPTED'`.
+  if (user != null && user.hasRoles('ADMIN') && statuses) {
+    comments = comments.where({
+      status: {
+        $in: statuses
```

Comment:
Ah, I see that this is not indeed a query, but appending a where clause. Makes sense. I like the pattern of composing a query one step at a time.